core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts

```diff
@@ -0,0 +1,419 @@
+    const answerIsCorrect = classificationResult.outcome.labelledAsCorrect;
+
+    let stateCard: StateCard;
+    let refreshInteraction: boolean = false;
+    let feedbackHtml: string = '';
+    let feedbackAudioTranslations: BindableVoiceovers = {};
+    let refresherExplorationId: string = '';
+    let missingPrerequisiteSkillId: string = '';
+    let remainOnCurrentCard: boolean = false;
+    let taggedSkillMisconceptionId: string = '';
+    let wasOldStateInitial: boolean = false;
+    let isFirstHit: boolean = true;
+    let isFinalQuestion: boolean = false;
+    let focusLabel: string;
+
+    this._numberOfAttemptedQuestions += 1;
+
+    if (answerIsCorrect) {
+      this._diagnosticTestCurrentTopicStatusModel.recordCorrectAttempt(
+        this._currentSkillId);
+    } else {
+      this._diagnosticTestCurrentTopicStatusModel.recordIncorrectAttempt(
+        this._currentSkillId);
+    }
+
+    this.getNextQuestionAsync().then((question: Question) => {
+      stateCard = this.createCard(question);
+      focusLabel = this._focusLabel;
+
+      successCallback(
+        stateCard, refreshInteraction, feedbackHtml,
+        feedbackAudioTranslations, refresherExplorationId,
+        missingPrerequisiteSkillId, remainOnCurrentCard,
+        taggedSkillMisconceptionId, wasOldStateInitial, isFirstHit,
+        isFinalQuestion, focusLabel
+      );
+      return answerIsCorrect;
+    }, () => {
+      // Test is finished.
+      const recommendedTopicIds: string[] = this.getRecommendedTopicIds();
+      this.diagnosticTestPlayerStatusService
+        .onDiagnosticTestSessionCompleted.emit(recommendedTopicIds);
+    });
+  }
+
+  getRecommendedTopicIds(): string[] {
+    let recommendedTopicIds: string[] = [];
+    let failedTopicIds: string[] = this.getFailedTopicIds();
+    let rootTopicIds: string[] = this.getRootTopicIds();
+    let failedRootTopicIds: string[] = rootTopicIds.filter(
+      topicId => failedTopicIds.indexOf(topicId) !== -1
+    );
+
+    if (failedRootTopicIds.length >= 2) {
+      recommendedTopicIds.push(failedRootTopicIds[0]);
+      recommendedTopicIds.push(failedRootTopicIds[1]);
+    } else if (failedRootTopicIds.length === 1) {
+      recommendedTopicIds.push(failedRootTopicIds[0]);
+    } else {
+      let sortedTopicIds = this.getSortedTopicIds();
+      for (let topicId of sortedTopicIds) {
+        if (failedTopicIds.indexOf(topicId) !== -1) {
+          recommendedTopicIds.push(topicId);
+          break;
+        }
+      }
+    }
+    return recommendedTopicIds;
+  }
+
+  getRootTopicIds(): string[] {
+    let topicIdToPrerequisiteTopicId = (
+      this._initialCopyOfTopicTrackerModel.getTopicIdToPrerequisiteTopicIds());
+    let rootTopicIds: string[] = [];
+
+    for (let topicId in topicIdToPrerequisiteTopicId) {
+      if (topicIdToPrerequisiteTopicId[topicId].length === 0) {
+        rootTopicIds.push(topicId);
+      }
+    }
+    return rootTopicIds;
+  }
+
+  getSortedTopicIds(): string[] {
```

Comment:
This could be more explicit -- I understand that we're getting a list of topic IDs that are sorted. But is this all topic IDs? visited ones? pending ones? etc. -- please explain better either in the variable name (preferred) or in a comment.

Reply:
Rephrased to `getSortedInitialTopicIds`.